doc/SONiC_YANG_Model_Guidelines.md

```diff
@@ -0,0 +1,270 @@
+.....
+.....
+}
+```
+
+### 2. Each leaf in YANG module should have same name as corresponding key-fields in ABNF.json.
+
+Example:
+Leaf names are same PACKET_ACTION, IP_TYPE and PRIORITY, which are .
+
+#### ABNF
+```
+        "NO-NSW-.....|.....": {
+            "PACKET_ACTION": "FORWARD",
+            "IP_TYPE": "IPv6ANY",
+            "PRIORITY": "955520",
+            .....
+            .....
+        },
+```
+
+####  YANG
+```
+            leaf PACKET_ACTION {
+                .....
+            }
+            leaf IP_TYPE {
+                .....
+            }
+            leaf PRIORITY {
+                .....
+            }
+```
+
+
+### 3. Data Node/Object Hierarchy of the an objects in YANG models will be same as for all the fields at same hierarchy in Config DB. If any exception is created then it must be recorded properly with comment under object level in YANG models. To see an example of a comment, please refer Guideline # 7.
+
+For Example:
+
+"Family" of VLAN_INTERFACE and "IP_TYPE" of ACL_RULE should be at same level in YANG model too.
+
+#### ABNF
+```
+"VLAN_INTERFACE": {
+        "Vlan100|2a04:f547:45:6709::1/64": {
+            "scope": "global",
+            "family": "IPv6"
+        }
+}
+"ACL_RULE": {
+        "NO-NSW-PACL-V4|DEFAULT_DENY": {
+            "PACKET_ACTION": "DROP",
+            "IP_TYPE": "IPv4ANY",
+            "PRIORITY": "0"
+        }
+}
+```
+####  YANG
+In YANG, "Family" of VLAN_INTERFACE and "IP_TYPE" of ACL_RULE is at same level.
+```
+container VLAN_INTERFACE {
+        description "VLAN_INTERFACE part of config_db.json";
+        list ..... {
+            ......
+            ......
+            leaf family {
+                type sonic-head:ip-family;
+            }
+        }
+}
+
+container ACL_RULE {
+        description "ACL_RULE part of config_db.json";
+        list ..... {
+            ......
+            ......
+            leaf IP_TYPE {
+                type sonic-head:ip_type;
+            }
+    }
+}
+```
+
+### 4.  If an object is part of primary-key in ABNF.json, then it should be a key in YANG model. In YANG models, a primary-key from ABNF.json can be represented either as name of a Container object or as a key field in List object. Exception must be recorded in YANG Model with a comment in object field. To see an example of a comment, please refer Guideline # 7.
+
+Example: VLAN_MEMBER dictionary in ABNF.json has both vlan-id and Port part of the key. So YANG model should have the same keys.
+
+#### ABNF
+```
+ "VLAN_MEMBER": {
+        "Vlan100|Ethernet0": {<<<< KEYS
+            "tagging_mode": "untagged"
+        }
+   }
+```
+#### YANG
+```
+container VLAN_MEMBER {
+    description "VLAN_MEMBER part of config_db.json";
+        list ..... {
+            key "vlanid port";<<<< KEYS
+       }
+}
+```
+
+### 5.  It is best to categorize YANG Modules based on a networking components. For Example, it is good to have separate modules for VLAN,  ACL, PORT and IP-ADDRESSES etc.
```

Comment:
Do we plan to model RPC's and Notifications? 
Can we put guidelines around rpc's and notifications?

Reply:
Please check latest PR https://github.com/Azure/SONiC/pull/464, step 23 and step 24.